src/hotspot/share/interpreter/abstractInterpreter.cpp

```diff
@@ -110,36 +110,57 @@ AbstractInterpreter::MethodKind AbstractInterpreter::method_kind(const methodHan
+  if (id != vmIntrinsics::_none) {
+    if (m->is_method_handle_intrinsic()) {
+      assert(MethodHandles::is_signature_polymorphic(id), "must match an intrinsic");
+      MethodKind kind = (MethodKind)(method_handle_invoke_FIRST +
+                                    vmIntrinsics::as_int(id) -
+                                    static_cast<int>(vmIntrinsics::FIRST_MH_SIG_POLY));
+      assert(kind <= method_handle_invoke_LAST, "parallel enum ranges");
+      return kind;
+    }
 
 #ifndef ZERO
-  switch (m->intrinsic_id()) {
-    // Use optimized stub code for CRC32 native methods.
-    case vmIntrinsics::_updateCRC32            : return java_util_zip_CRC32_update;
-    case vmIntrinsics::_updateBytesCRC32       : return java_util_zip_CRC32_updateBytes;
-    case vmIntrinsics::_updateByteBufferCRC32  : return java_util_zip_CRC32_updateByteBuffer;
-    // Use optimized stub code for CRC32C methods.
-    case vmIntrinsics::_updateBytesCRC32C             : return java_util_zip_CRC32C_updateBytes;
-    case vmIntrinsics::_updateDirectByteBufferCRC32C  : return java_util_zip_CRC32C_updateDirectByteBuffer;
-    case vmIntrinsics::_intBitsToFloat:      return java_lang_Float_intBitsToFloat;
-    case vmIntrinsics::_floatToRawIntBits:   return java_lang_Float_floatToRawIntBits;
-    case vmIntrinsics::_longBitsToDouble:    return java_lang_Double_longBitsToDouble;
-    case vmIntrinsics::_doubleToRawLongBits: return java_lang_Double_doubleToRawLongBits;
-    default:                                 break;
-  }
+    switch (m->intrinsic_id()) {
+      // Use optimized stub code for CRC32 native methods.
+      case vmIntrinsics::_updateCRC32: return java_util_zip_CRC32_update;
+      case vmIntrinsics::_updateBytesCRC32: return java_util_zip_CRC32_updateBytes;
+      case vmIntrinsics::_updateByteBufferCRC32: return java_util_zip_CRC32_updateByteBuffer;
+      // Use optimized stub code for CRC32C methods.
+      case vmIntrinsics::_updateBytesCRC32C: return java_util_zip_CRC32C_updateBytes;
+      case vmIntrinsics::_updateDirectByteBufferCRC32C: return java_util_zip_CRC32C_updateDirectByteBuffer;
+      case vmIntrinsics::_intBitsToFloat: return java_lang_Float_intBitsToFloat;
+      case vmIntrinsics::_floatToRawIntBits: return java_lang_Float_floatToRawIntBits;
+      case vmIntrinsics::_longBitsToDouble: return java_lang_Double_longBitsToDouble;
+      case vmIntrinsics::_doubleToRawLongBits: return java_lang_Double_doubleToRawLongBits;
+      case vmIntrinsics::_dsin: return java_lang_math_sin;
+      case vmIntrinsics::_dcos: return java_lang_math_cos;
+      case vmIntrinsics::_dtan: return java_lang_math_tan;
+      case vmIntrinsics::_dabs: return java_lang_math_abs;
+      // _dsqrt will be selected for both Math::sqrt and StrictMath::sqrt, but the latter
+      // is native. Keep treating it like a native method in the interpreter
+      case vmIntrinsics::_dsqrt: return m->is_native() ? native : java_lang_math_sqrt;
```

Comment:
Maybe we should add an assert that m must be one of the two methods mentioned by the above comment?

Reply:
Ok. I'll add an assertion that allows for the StrictMath variant to be re-implemented in java